Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp

```diff
@@ -2904,16 +2960,78 @@ class LowerToAir {
             }
         }
 
-        if (isX86()) {
-            Air::Opcode casOpcode = OPCODE_FOR_WIDTH(BranchAtomicStrongCAS, atomic->accessWidth());
-            reloopBlock->append(relaxedMoveForType(atomic->type()), m_value, oldValue, m_eax);
-            reloopBlock->append(trappingInst(m_value, casOpcode, m_value, Arg::statusCond(MacroAssembler::Success), m_eax, newValue, address));
+        Tmp boolResult = m_code.newTmp(GP);
+        reloopBlock->append(trappingInst(m_value, storeCondOpcode(atomic->accessWidth()), m_value, newValue, address, boolResult));
+        reloopBlock->append(BranchTest32, m_value, Arg::resCond(MacroAssembler::Zero), boolResult, boolResult);
+        reloopBlock->setSuccessors(doneBlock, reloopBlock);
+    }
+
+    void appendGeneralAtomic64(Air::Opcode opcode)
+    {
+        using namespace Air;
+        AtomicValue* atomic = m_value->as<AtomicValue>();
+
+        Arg address = addr(m_value);
+        auto [ operandLo, operandHi ] = tmpsForInt64(atomic->child(0));
+        auto [ oldValueLo, oldValueHi] = tmpsForInt64(atomic);
+        Tmp newValueLo, newValueHi;
+        if (opcode == Air::Nop) {
+            newValueLo = operandLo;
+            newValueHi = operandHi;
         } else {
-            RELEASE_ASSERT(isARM64());
-            Tmp boolResult = m_code.newTmp(GP);
-            reloopBlock->append(trappingInst(m_value, storeCondOpcode(atomic->accessWidth(), atomic->hasFence()), m_value, newValue, address, boolResult));
-            reloopBlock->append(BranchTest32, m_value, Arg::resCond(MacroAssembler::Zero), boolResult, boolResult);
+            newValueLo = tmpForType(Int32);
+            newValueHi = tmpForType(Int32);
         }
+
+        // We need a CAS loop or a LL/SC loop. Using prepare/attempt jargon, we want:
+        //
+        // Block #reloop:
+        //     Prepare
```

Comment:
I guess this is a general issue, but a tight loop like this will break the JSC watchdog, which might make fuzzing harder. There are other places that are affected too, but we should keep this section in mind if we ever get around to fixing that.